module/zfs/metaslab.c

```diff
@@ -1420,8 +1420,15 @@ metaslab_load(metaslab_t *msp)
 	boolean_t success = B_FALSE;
 
 	ASSERT(MUTEX_HELD(&msp->ms_lock));
-	ASSERT(!msp->ms_loaded);
-	ASSERT(!msp->ms_loading);
+
+	/*
+	 * There may be another thread loading the same metaslab, if that's
+	 * the case just wait until the other thread is done and return.
+	 */
+	metaslab_load_wait(msp);
```

Comment:
I'm totally fine with the code as-is, but it would be nice if it was easier to trace "real loads" (e.g. with dtrace / bpftrace).  This change makes that a bit harder since calls to metaslab_load() may be a no-op, may have to wait for an in-progress load to complete, or may do the load itself.  It might make sense to either move the "really loading" code into its own function (e.g. `metaslab_load_impl()`), or add SDT probes for the cases I mentioned above?

Reply:
That makes sense. I'll factor the real loading in its own function.